docs/openid.md

```diff
@@ -590,3 +590,42 @@ oidc_providers:
         display_name_template: "{{ user.first_name }} {{ user.last_name }}"
         email_template: "{{ user.email }}"
 ```
+
+### Mastodon / Fediverse
+
+Single-Sign On provided by [Mastodon](https://docs.joinmastodon.org/) and other Fediverse websites work with Synapse, provided they have implemented the [Mastodon OAuth API](https://docs.joinmastodon.org/spec/oauth/).
+
+This example assumes: 
+* the Mastodon instance website URL is `https://your.mastodon.instance.url/` 
+* you want to create an App titled *my_synapse_app*
+
+The first step is to create an App using the Mastodon's [Create an application API](https://docs.joinmastodon.org/methods/apps/#create). There are several ways to do this but in the example below we are using CURL.
+
+Send the following request:
+```sh
+curl -d "client_name=my_synapse_app&redirect_uris=https://[synapse_public_baseurl]/_synapse/client/oidc/callback" -X POST https://your.mastodon.instance.url/api/v1/apps
+```
+
+You should get the following response, and you should write down:
+```json
+{"client_id":"someclientid_123","client_secret":"someclientsecret_123","id":"12345","name":"my_synapse_app","redirect_uri":"https://[synapse_public_baseurl]/_synapse/client/oidc/callback","website":null,"vapid_key":"somerandomvapidkey_123"}
+```
+
+As the Synapse login mechanism needs an attribute to uniquely identify users, and that endpoint does not return a sub property, an alternative subject_claim has to be set. Your Synapse configuration should be the following:
+
+```yaml
+oidc_providers:
+  - idp_id: fediverse
+    idp_name: "Fediverse Example"
+    discover: false
+    issuer: "https://your.mastodon.instance.url/@admin"
```

Comment:
Is the name `@admin` always correct, or can this depend on the how the mastodon instance is configured?

Reply:
I actually wonder if the `issuer` is used anywhere with this particular auth flow. Since Mastodon auth is undocumented currently, I got the example config from a foreign website and the example config had `https://instance/@admin` so I just went with it, and it worked with my setup too. 

When I tried removing `issuer` from Synapse config, homeserver refused to start, stating `'issuer' is a required property`. I looked into [Husky Android app](https://git.sr.ht/~captainepoch/husky) source and I can't see anything related to issuer (even when doing "grep -ri "issuer") . 

```kotlin
private fun redirectUserToAuthorizeAndLogin(domain: String, clientId: String) {
    /* To authorize this app and log in it's necessary to redirect to the domain given,
     * login there, and the server will redirect back to the app with its response. */
    val endpoint = MastodonApi.ENDPOINT_AUTHORIZE
    val parameters = mapOf(
        "client_id" to clientId,
        "redirect_uri" to oauthRedirectUri,
        "response_type" to "code",
        "scope" to OAUTH_SCOPES
    )
    val url = "https://" + domain + endpoint + "?" + toQueryString(parameters)
    val uri = Uri.parse(url)
    if (!openInCustomTab(uri, this)) {
        val viewIntent = Intent(Intent.ACTION_VIEW, uri)
        if (viewIntent.resolveActivity(packageManager) != null) {
            startActivity(viewIntent)
        } else {
            binding.domainEditText.error = getString(R.string.error_no_web_browser_found)
            setLoading(false)
        }
    }
}
```

Reply:
https://github.com/matrix-org/synapse/blob/cc3a52b33df72bb4230367536b924a6d1f510d36/synapse/config/oidc.py#L340-L342

So this is only used to "validate tokens". Is it used?

- Not used [here](https://github.com/matrix-org/synapse/blob/09de2aecb05cb46e0513396e2675b24c8beedb68/synapse/handlers/oidc.py#L389-L398) because we already have a client secret configured
- Used in various places if backchannel logout is enabled---but it isn't configured, and [defaults to false](https://github.com/matrix-org/synapse/blob/cc3a52b33df72bb4230367536b924a6d1f510d36/synapse/config/oidc.py#L297).
- Not used [here](https://github.com/matrix-org/synapse/blob/09de2aecb05cb46e0513396e2675b24c8beedb68/synapse/handlers/oidc.py#L586-L590) because discovery is disabled

So I don't think it's used. Let's leave it as it is.